File: workspace_interpreter/expr.py
"""Public API for the expression evaluator.

Parsed ASTs are cached per source string in a process-wide dict:
re-evaluating the same expression (e.g. a [bind:] clause inside a
216-iteration foreach) skips the tokenize+parse step entirely. The
cache is unbounded — fine for workspace YAML, where the set of
distinct expression strings is finite and bounded by the spec.

Diagnostics: parse failures are logged at WARNING; non-empty
expressions that evaluate to null are logged at DEBUG (legit for
e.g. $selection.fill with nothing selected, but also flags typos
during development when debug logging is enabled).
"""

from __future__ import annotations
import logging
import re

from workspace_interpreter.expr_types import Value, ValueType
from workspace_interpreter.expr_parser import parse, ParseError
from workspace_interpreter.expr_eval import eval_node


_INTERP_RE = re.compile(r"\{\{(.+?)\}\}")

# Cache of parsed ASTs keyed by source string. ``None`` cached for
# unparseable input so we don't reparse known-bad strings.
_AST_CACHE: dict = {}

_log = logging.getLogger(__name__)


def evaluate(expr_str: str, ctx: dict) -> Value:
    """Evaluate an expression string in expression context (no {{}}).

    Returns a typed Value. Never raises — returns Value.null() on error.
    """
    if not expr_str or not isinstance(expr_str, str):
        return Value.null()
    source = expr_str.strip()
    if source in _AST_CACHE:
        ast = _AST_CACHE[source]
    else:
        try:
            ast = parse(source)
        except (ParseError, Exception) as e:
            _log.warning("parse failed: %r: %s", source, e)
            ast = None
        _AST_CACHE[source] = ast
    if ast is None:
        return Value.null()
    try:
        result = eval_node(ast, ctx)
    except Exception as e:
        _log.warning("eval raised: %r: %s", source, e)
        return Value.null()
    if result.type == ValueType.NULL:
        _log.debug("null result: %r", source)
    return result


def evaluate_text(text: str, ctx: dict) -> str:
    """Evaluate a text string with embedded {{expr}} regions.

    Returns the string with each {{expr}} replaced by its evaluated
    value coerced to a string. Text outside {{}} is literal.
    """
    if not text or not isinstance(text, str) or "{{" not in text:
        return text if isinstance(text, str) else ""

    def _replace(match):
        expr_str = match.group(1).strip()
        val = evaluate(expr_str, ctx)
        return val.to_string()

    return _INTERP_RE.sub(_replace, text)

File: jas_flask/static/js/engine/expr.mjs
// Public API for the JS expression evaluator.
//
// Parallels `workspace_interpreter/expr.py`. Parsed ASTs are cached
// per source string in a process-wide Map — re-evaluating the same
// expression (inside a foreach or a per-frame overlay) skips the
// tokenize + parse step.
//
// Returns a typed Value (value.mjs). Never throws — returns mkNull()
// on any failure path.

import { parse, ParseError } from "./parser.mjs";
import { evalNode } from "./evaluator.mjs";
import { mkNull, toStringCoerce } from "./value.mjs";
import { Scope } from "./scope.mjs";

// Interpolation regex matches `{{…}}` blocks with minimal capture.
const INTERP_RE = /\{\{(.+?)\}\}/g;

// AST cache keyed by source string. `null` caches unparseable input
// so known-bad strings don't re-tokenize.
const AST_CACHE = new Map();

/**
 * Evaluate an expression string against a Scope. Returns a Value;
 * null/empty input yields mkNull().
 *
 * @param {string} source - the expression source
 * @param {Scope|object} scopeOrCtx - a Scope, or a plain dict that
 *   will be wrapped in a fresh Scope
 */
export function evaluate(source, scopeOrCtx) {
  if (!source || typeof source !== "string") return mkNull();
  const trimmed = source.trim();
  if (trimmed.length === 0) return mkNull();

  let ast;
  if (AST_CACHE.has(trimmed)) {
    ast = AST_CACHE.get(trimmed);
  } else {
    try {
      ast = parse(trimmed);
    } catch (e) {
      if (e instanceof ParseError) {
        if (isDebugEnabled()) {
          console.warn(`expr parse failed: ${JSON.stringify(trimmed)}: ${e.message}`);
        }
        ast = null;
      } else {
        throw e;
      }
    }
    AST_CACHE.set(trimmed, ast);
  }

  if (ast === null) return mkNull();

  const scope = scopeOrCtx instanceof Scope ? scopeOrCtx : new Scope(scopeOrCtx);
  try {
    const result = evalNode(ast, scope);
    if (result.kind === "null" && isDebugEnabled()) {
      console.debug(`expr null result: ${JSON.stringify(trimmed)}`);
    }
    return result;
  } catch (e) {
    if (isDebugEnabled()) {
      console.warn(`expr eval raised: ${JSON.stringify(trimmed)}: ${e.message}`);
    }
    return mkNull();
  }
}

/**
 * Evaluate a text string with embedded `{{…}}` expression regions.
 * Text outside the braces is literal. Each evaluated expression is
 * coerced to a string via `toStringCoerce`.
 */
export function evaluateText(text, scopeOrCtx) {
  if (!text || typeof text !== "string") return "";
  if (!text.includes("{{")) return text;
  const scope = scopeOrCtx instanceof Scope ? scopeOrCtx : new Scope(scopeOrCtx);
  return text.replace(INTERP_RE, (_, expr) =>
    toStringCoerce(evaluate(expr.trim(), scope))
  );
}

/**
 * Clear the AST cache. Useful between tests or when hot-reloading
 * YAML during development.
 */
export function clearCache() {
  AST_CACHE.clear();
}

// Debug-mode logging gated by the JAS_DEBUG_EXPR env var (Node) or a
// window.JAS_DEBUG_EXPR boolean (browser). Matches the gate used in
// the OCaml and Swift interpreters.
let _debugCached = null;
function isDebugEnabled() {
  if (_debugCached !== null) return _debugCached;
  if (typeof process !== "undefined" && process.env) {
    _debugCached = process.env.JAS_DEBUG_EXPR === "1";
    return _debugCached;
  }
  if (typeof globalThis !== "undefined" && globalThis.JAS_DEBUG_EXPR) {
    _debugCached = true;
    return true;
  }
  _debugCached = false;
  return false;
}
